### DISCUSSION TITLE
Using cheap servo testers as the throttle/running engine without Rc transmitter and receiver

### DISCUSSION THREAD

**Discussion by DG-PROPULSION** [Jehanzeb1973/Ardu_ECU#4] [Ideas]
I have been trying to mimic the RC_mode_pin and RC_throttle_pin signals by using the servo testers hooked to GPIO 13 and GPIO 35 respectively, as they also output the same PWM pulse as the Rc receiver outputs.

But ECU is still giving me error code 8 which is RC signal lost.
My idea is to bench test the engine without needing a Rc transmitter and receiver and just using a servo tester as throttle.

**Comment by Jehanzeb1973** [comment]
Did you ensure a common ground between servo tester and ESP32. I suspect this is a hardware issue since the code has been tested already. Can you connect an oscilloscope and check the voltage levels from servo tester. The circuit should work with both 5v and 3.3v as voltage is limited by a zener connected between GPIO and ground. Having common ground is important. I have burned a few boards when i forgot to connect all grounds together

**Reply by DG-PROPULSION** [reply]
Yup I did.
And I tested with 3 different servo testers. I didn't connect them to the oscilloscope BUT i tested all of my servo testers with an ESC and a bldc motor and motor runs fine. So, servo testers are definitely producing a pwm pulse.

**Reply by DG-PROPULSION** [reply]
I have tried everything. Could not even reset the error code "8".

**Reply by Jehanzeb1973** [reply]
Resetting the error is built into firmware. If you hold down the button for 5 seconds, it should reset the error code. If the system does not detect RC signl, it will again show the error. What about Firmware Rev 9. Does it detect the servo signal

**Reply by DG-PROPULSION** [reply]
I tested with both firmware 9 and 10 but same result.
Cannot get it past error "8"

**Reply by Jehanzeb1973** [reply]
I am running it with Futaba Radio and REceiver. I will test it with Servo Tester and get back to you

**Reply by DG-PROPULSION** [reply]
ok.
Thanks

**Reply by DG-PROPULSION** [reply]
Hi @Jehanzeb1973 
Did you get a chance to use the servo testers?
Thanks

**Comment by DG-PROPULSION** [comment]
Hi @Jehanzeb1973 
I have also updated your schematic to something that can be easily printed on a pcb. My plan is to eventually get some PCBs manufactured and try to run them.
In my design the rc_mode_signal and rc_throttle servo testers are getting their power from +3.3v pin of the esp32.

I have also tried to keep all the separate PCBs (hall sensor pcb, optocoupler PCB etc.) out of the main board just in case something fails and we may have to replace it.
Let me know how I can share the updated schematic here?

I also noticed in your second video that you changed the hall effect module BUT your schematic still shows YS-27. Which one are you currently using?

**Comment by Jehanzeb1973** [comment]
You can send me schematic at jehanzeb@digipak.org. Regarding Servo Tester Power up, Did you check if it runs with 3.3V as this is something i was planning to check. It could be that Servo Tester only works at 5V.
Now for Hall Sensor, this is a tricky one. Initially i used YS-27 which requires adjusting potentiometer manually. If it is not setup properly, it will either not read the RPM or give false counts. I had to play with it almost every time and then i was not sure if the RPM is accurate or is it reading higher than actual.
Then i moved to OpAmp based RPM conditioner. I have only two analogue RPM sensors from Xicoy and both of them have different noise and signal levels. The opamp circuit for one sensor does not work with the other sensor. However once i have a OPAmp amplifier made for one sensor, it is rock stable and doesnt give any false readings so now i am using opAmp Amplifier as described in second video. For the sensors that i have looks like the opAmp amplifier is specific to that sensor. May be the sensors are from different batches. This is the problem i will address now. All these sensors are based on NXP geomagnetic sensors of KMZ series. I plan to look at these sensors and come to a conclusion as having custom amplifier for each sensor is not a good way to address this issue

**Reply by DG-PROPULSION** [reply]
Hi @Jehanzeb1973 
When I was trying with the servo testers and getting “no rc signal), I tried running them off of both 5v (from ESC BEC) and from 3.3v (ESP32s) supply.

The lights on the servo testers were ON in both cases though.
From what I have seen and read the logic level for the PWM signals is fairly low at about ~2.2v above which is considered HIGH.
But I will still wait for your testing before finalising my schematic. The only issue is we don’t have 5v supply on the board so if we need 5v for the servo testers then it has to be externally supplied or maybe using a separate 12v to 5v buck converter like we use for the Hall effect sensor.

The Op-Amp amplifier you are using is an off the shelf unit or self made?

Thanks

**Reply by Jehanzeb1973** [reply]
Hi @dstromeng, I have tested with Servo tester and it works fine. The Futaba receiver is powered from BEC on one speed controller. Switching to Servo Tester doesnt change supply voltage and is 5V in both cases. I am not using 3.3V to power Futaba Receiver or Servo Tester. 
For OpAmp Circuit, i have used generic LM358 OpAmp and used the following site to calculate the resistor values . The circuit i have used is http://hyperphysics.phy-astr.gsu.edu/hbase/Electronic/schmitt.html. I made a video but its too large and cannot attach to this post

**Reply by DG-PROPULSION** [reply]
Hi @Jehanzeb1973 
So you had to use 2 servo testers right?
one for rc mode signal and one for rc throttle signal?

**Reply by DG-PROPULSION** [reply]
I have emailed you the schematic. Let me know how does it look to you.

You will be able to see I am taking the supply for 2 servo testers from one of the BECs.

One thing that confused me is one of the BEC supply voltage is being fed through (5v to 3.3v logic level converter) to the VCC+ pin of the ESP32. 
I thought that was to supply power to the ESP32 but it didn’t powered my ESP32. So I kept connecting that pin and had to power my ESP32 using USB port.

Furthermore I connected 1 servo testers PWM/signal pin directly to the throttle input pin of the ESP32 without 1k resistor and diode and other servo tester directly to RC mode pin of ESP32 just to see what would happen.
Nothing happened. ESP32 was fine but still did not get past failure mode “8”.

**Reply by DG-PROPULSION** [reply]
Hi @Jehanzeb1973 
Will you be able to explain this Op-Amp circuit and its working? I ama noob trying to understand a lot of things together. I watched a few videos online and got confused even more. How you arrived at the resistance values, what's the voltage coming out of the rpm sensor and how to convert it to a proper 0-5v digital pulse?

Also do you know what type of hall effect sensor is used by rc turbines?

Thanks.

**Reply by DG-PROPULSION** [reply]
Hi @Jehanzeb1973 
I tested my servo testers (5 of them) using a handheld multimeter cum oscilloscope. All Of them were fed from speed Controllers BEC (the voltage input was 5v)
Now I wasn’t sure if PWM output signal should be measured as Vdc or Vac on the multimeter so I tried both. 
when measured as Vdc the peak to peak voltage is just 3volts BUT when measured as Vac the peak to peak voltage of the PWM signal was 4.2volts which seemed more like it should be.

Now feeding that PWM output into the throttle or rc mode pin still didn’t clear the error.

Do you think this error is persistent because I haven’t connected all inputs and outputs to the ecu yet? 
currently I only have a thermocouple module and 2 screens connected to the ECU and I am using Firmware v9 (to support the screens). Then I just tried connecting these servo testers to see if throttle percentage will rise on the screen but it never got past that error.

regards

**Comment by Jehanzeb1973** [comment]
Hi @dstromeng , 
No. I had only one tester so i used it on Throttle input. The mode was left unconnected and That worked fine. I could control starter RPM using Servo Tester

**Reply by DG-PROPULSION** [reply]
Which mode were you testing your engine at?
I thought you couldn’t go past the abort(all) command without connecting rc mode

**Comment by Jehanzeb1973** [comment]
It was in the normal mode which is default at startup. In this mode starter runs from throttle control. You cannot start the engine however as you need mode switch to start the engine. Also RC signl error is also trigerred from throttle signal so connecting to throttle only eliminates "RC Signal lost" error

**Reply by DG-PROPULSION** [reply]
Ok.
That’s exactly what I was doing. Though I hadn’t had the starter connected. BUT I was reading the throttle % from the screen.
But like I said it wasn’t going past the error 8

**Comment by Jehanzeb1973** [comment]
I got two of these sensors from xicoy. When you power them up with 5v, the signal output without any rotation is approximately 2.5v (one was reading 2.6v and other 2.5v). When you start the rotation you get a sinusoidal signal One sensor was giving only a signal level of 0.7 volts while the other was giving a signal level of  1 volt. Also there was a noise on these output of 0.05-0.01 volts.We want a square wave at the RPM sensor input and a voltage level as per ESP32 requirements Theoretically: LOW LEVEL range is -0.3V to 0.825V HIGH LEVEL range is 2.475V to 3.6V. The opamp circuit takes this sinewave and converts it into square wave and ignores the noise. The resistor values will be as per your noise level and center value.
Both these sensor are using NXP’s KMZ10 or 10a geomagnetic sensor
Regards

Jehanzeb Khan


http://DigiPak.org/

**Reply by DG-PROPULSION** [reply]
Hi @Jehanzeb1973 
I am curious how the rc jet engines are able to get the reliable RPMs using the same or similar Xicoy sensors?
Do they use Op-amps as well somewhere in their ECU circuit. If that’s the case we should add the Op-amp to the ECU schematic as well.

Regards.

**Comment by Jehanzeb1973** [comment]
Hi @dstromeng
The issue is that unlike the other sensors, the RPM sensors I have, are giving different voltages and noise so it is difficult to generalize. It could be that one of the sensors is bad but I have no way to tell as they are measuring RPM. Just the signal and noise levels are different.
I am working on this and have ordered some sensor ic’s so may be I can come to a conclusion. Once I have the answer, I will probably do another explainer video on this.
On the issue of Schematic, I would still leave RPM sensor circuit out of it to keep this project flexible. AS far as the ECU is concerned, it needs a pulse signal on RPM pin for each Revolution. It is not bound to sensor type.. The sensor can be a magnetic RPM pickup, A hole through the nut optical sensor or a reflective optical sensor or any other type.
Regards

Jehanzeb Khan
DIGIPAK Pvt. Limited


http://DigiPak.org/

**Reply by DG-PROPULSION** [reply]
I have a couple Rc jet engine ECUs. All of them have the Op-amps right next to the rpm signal input. I will be checking with a multimeter later today to see if the rpm signal is going to the Op-amp.

**Comment by Jehanzeb1973** [comment]
Hi @dstromeng
The servo signal is going to GPIO through a resistor and then voltage limited using a Zener diode. How did you connect the diodes between GPIO pin and ground ? These are 3.3v Zener diodes which need to be reversed biased so the side with black band needs to be towards GPIO and other side is to ground.

You will not read input signal if you have connected these in forward bias.
Regards


Jehanzeb Khan
DIGIPAK Pvt. Limited


http://DigiPak.org/

**Reply by DG-PROPULSION** [reply]
I connected the servo pwm signal both directly(without resistor and diode) to the GPIO pin and then with a resistor and 3.3v zener diode reverse biased.
But more often it was connected directly as I read somewhere ESP32 can take 5v pwm signals at GPIO pins without damaging it.

the ESP32 has been working fine after all this. Atleast that’s what it seems like. It runs the code, creates a webserver etc.

**Comment by Jehanzeb1973** [comment]
I have never tried 5v to ESP32 pins as I have heard it can damage ESP32. I have only tried 3.3v . If you are still not getting Servo signal to ESP32. Try feeding the Servo signal that ESP is generating back to its input pin i..e The Fuel pump control or starter motor control back to ESP32. Don’t use any adapter in-between as it is from ESP32pin to ESP32 Pin
REgards

Jehanzeb Khan
DIGIPAK Pvt. Limited


http://DigiPak.org/

**Reply by DG-PROPULSION** [reply]
Hi @Jehanzeb1973 
I tried feeding the servo signal (from both starter speed control and fuel pump speed control) back to both rc mode pin and throttle pin BUT to no avail.

still won’t go past RC signal lost error 8.

what everything in the code causes this error 8. I am out of options now.

**Reply by DG-PROPULSION** [reply]
Also do you think installing a wrong library may have caused this?
Since a lot of libraries were included in the firmware package I had to search online and install them.

**Reply by DG-PROPULSION** [reply]
Hi @Jehanzeb1973 
I am not sure if it's doable here BUT it would be a good idea to include the libraries with the code where you have downloaded them from. I think there are a lot of options for newbies like me to download a corrupt library.

**Comment by DG-PROPULSION** [comment]
Hi @Jehanzeb1973 
I was able to successfully test the servo tester with the ECU.
Turns out the libraries that I may have downloaded were corrupt or wrong. I didn't realize that you have to install the important libraries from inside the Arduino IDE software. I downloaded some from the internet and installed them which wasn't the correct way of doing it. I deleted all the libraries and reinstalled them from inside the arduino IDE and the ECU started working as it should.

Thanks.

**Comment by Jehanzeb1973** [comment]
I cannot copy the libraries as these are too large. I will try to copy links of libraries. Although I did that for max6675

Jehanzeb Khan
DIGIPAK Pvt. Limited



http://DigiPak.org/

**Comment by Jehanzeb1973** [comment]
It is possible but generally wrong library will have different variable declarations and will give you errors at compile time

Jehanzeb Khan
DIGIPAK Pvt. Limited


http://DigiPak.org/

**Comment by Jehanzeb1973** [comment]
Hi @dstromeng
Great to know that the issue has been resolved. Let me know how the testing of the engine goes.
Regards
Jehanzeb Khan
DIGIPAK Pvt. Limited


http://DigiPak.org/
